lib/core/src/sync/client.rs

```diff
@@ -0,0 +1,71 @@
+use anyhow::{anyhow, Result};
+
+use async_trait::async_trait;
+use tokio::sync::Mutex;
+
+use super::model::sync::{
+    syncer_client::SyncerClient as ProtoSyncerClient, ListChangesReply, ListChangesRequest,
+    ListenChangesRequest, Record, SetRecordReply, SetRecordRequest,
+};
+
+#[async_trait]
+pub(crate) trait SyncerClient: Send + Sync {
+    async fn connect(&self, connect_url: String) -> Result<()>;
+    async fn set_record(&self, req: SetRecordRequest) -> Result<SetRecordReply>;
+    async fn list_changes(&self, req: ListChangesRequest) -> Result<ListChangesReply>;
+    async fn listen_changes(
+        &self,
+        req: ListenChangesRequest,
+    ) -> anyhow::Result<tonic::codec::Streaming<Record>>;
+    async fn disconnect(&self) -> Result<()>;
+}
+
+pub(crate) struct BreezSyncerClient {
+    inner: Mutex<Option<ProtoSyncerClient<tonic::transport::Channel>>>,
+}
+
+impl BreezSyncerClient {
+    pub(crate) fn new() -> Self {
+        Self {
+            inner: Default::default(),
+        }
+    }
+}
+
+#[async_trait]
+impl SyncerClient for BreezSyncerClient {
+    async fn connect(&self, connect_url: String) -> Result<()> {
+        let mut client = self.inner.lock().await;
+        *client = Some(ProtoSyncerClient::connect(connect_url).await?);
+        Ok(())
+    }
+
+    async fn set_record(&self, req: SetRecordRequest) -> Result<SetRecordReply> {
+        let Some(mut client) = self.inner.lock().await.clone() else {
```

Comment:
This will lock the client until the end of the invocation. I think it is better to extract out getting the client to its own function: `get_client()` so the lock is released right after the get.

Reply:
The inner data (client) is cloned and the lock is dropped there